File: cip/src/formats/mobi_reader.cpp
// Copyright (c) 2022 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by GNU General Public License
// that can be found in the LICENSE file.

#include "formats/mobi_reader.h"

#include <QDebug>

MobiReader::MobiReader(QObject* parent) : QObject(parent) {

}

MobiReader::~MobiReader() {
  this->cleanup();
}

void MobiReader::cleanup() {
  if (rawml_ != nullptr) {
    mobi_free_rawml(rawml_);
    rawml_ = nullptr;
  }

  if (mobi_ != nullptr) {
    mobi_free(mobi_);
    mobi_ = nullptr;
  }
}

bool MobiReader::load(const QString& filepath) {
  this->cleanup();
  mobi_ = mobi_init();
  if (mobi_ == nullptr) {
    qWarning() << "Failed to init mobi instance";
    return false;
  }

  MOBI_RET ret = mobi_load_filename(mobi_, filepath.toStdString().c_str());
  if (ret != MOBI_SUCCESS) {
    qWarning() << "Failed to load mobi file:" << filepath << ", ret:" << ret;
    return false;
  }

  rawml_ = mobi_init_rawml(mobi_);
  if (rawml_ == nullptr) {
    qWarning() << "Failed to init rawml:" << filepath;
    return false;
  }

  return true;
}

int MobiReader::numPages() const {
  if (mobi_ == nullptr || rawml_ == nullptr) {
    return -1;
  }

  const MOBI_RET ret = mobi_parse_rawml(rawml_, mobi_);
  if (ret != MOBI_SUCCESS) {
    qWarning() << "Failed to parse rawml in mobi file, ret:" << ret;
    return -1;
  }

  size_t uid = -1;
  while (true) {
    uid += 1;
    MOBIPart* part = mobi_get_part_by_uid(rawml_, uid);
    if (part == nullptr) {
      break;
    }
  }

  return static_cast<int>(uid);
}

bool MobiReader::readPage(int number, QString& text) {
  if (mobi_ == nullptr || rawml_ == nullptr) {
    qWarning() << "mobi file not load yet!";
    return -1;
  }

  const MOBI_RET ret = mobi_parse_rawml(rawml_, mobi_);
  if (ret != MOBI_SUCCESS) {
    qWarning() << "Failed to parse rawml in mobi file, ret:" << ret;
    return -1;
  }

  int uid = -1;
  MOBIPart* part = nullptr;
  while (uid < number) {
    uid += 1;
    part = mobi_get_part_by_uid(rawml_, uid);
    if (part == nullptr) {
      break;
    }
  }

  if (part == nullptr) {
    qWarning() << "part is null";
    return false;
  }

  const char* data = reinterpret_cast<char*>(part->data);
  text = data;

  return true;
}

File: cip/src/formats/mobi_reader.rs
// Copyright (c) 2023 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by GNU General Public License
// that can be found in the LICENSE file.

use encoding_rs::WINDOWS_1252;
use mobi::headers::TextEncoding;
use mobi::Mobi;
use std::path::Path;

use crate::error::{Error, ErrorKind};

pub struct MobiReader {
    doc: Mobi,
}

impl MobiReader {
    /// Open an mobi file.
    ///
    /// # Errors
    /// Returns error if mobi file does not found or is invalid.
    pub fn open<P: AsRef<Path>>(path: P) -> Result<Self, Error> {
        let doc = Mobi::from_path(path)?;
        Ok(Self { doc })
    }

    #[must_use]
    pub fn pages(&self) -> usize {
        self.doc.readable_records_range().len()
    }

    /// Read specific page.
    ///
    /// # Errors
    /// Returns error if failed to read file content.
    pub fn read_page(&self, page: usize) -> Result<String, Error> {
        let encoding = self.doc.text_encoding();
        let records = self.doc.raw_records();
        let records = records.records();
        if page < records.len() {
            let record = records[page];
            let content = record.content;
            match encoding {
                TextEncoding::UTF8 | TextEncoding::Unknown(_) => {
                    Ok(String::from_utf8_lossy(content).into_owned())
                }
                TextEncoding::CP1252 => match WINDOWS_1252.decode_with_bom_removal(content) {
                    (dest, true) => Ok(dest.to_string()),
                    (_, false) => Err(Error::new(ErrorKind::InvalidMobiPage, "Encode error")),
                },
            }
        } else {
            Err(Error::new(ErrorKind::InvalidMobiPage, "Invalid page index"))
        }
    }
}
